File: mainwindow.cpp
#include "mainwindow.h"
#include "ui_mainwindow.h"

#include "blusage.h"
#include "accountdialog.h"

#include <qnetworkrequest.h>
#include <qnetworkreply.h>
#include <qmessagebox.h>

MainWindow::MainWindow(QWidget *parent) :
    QMainWindow(parent),
    ui(new Ui::MainWindow)
{
    ui->setupUi(this);

    this->setWindowIcon(QIcon(":/blusage.png"));

    QMenu *menuFile = new QMenu(QString("File"), this);
    menuFile->addAction(ui->actionAccount);
    menuFile->addAction(ui->actionUpdate);
    menuFile->addSeparator();
    menuFile->addAction(ui->actionQuit);

    QMenu *menuHelp = new QMenu("Help", this);
    menuHelp->addAction(ui->actionAbout);

    ui->menuBar->addMenu(menuFile);
    ui->menuBar->addMenu(menuHelp);

    this->connect(ui->actionAccount, SIGNAL(triggered()), SLOT(showAccountEditor()));
    this->connect(ui->actionUpdate, SIGNAL(triggered()), SLOT(updateUsage()));
    this->connect(ui->actionAbout, SIGNAL(triggered()), SLOT(about()));

    networkAccessManager = new QNetworkAccessManager;
    this->connect(networkAccessManager, SIGNAL(finished(QNetworkReply*)), SLOT(fetchedUsages(QNetworkReply*)));
    this->connect(networkAccessManager, SIGNAL(sslErrors(QNetworkReply*,QList<QSslError>)), SLOT(allowConnection(QNetworkReply*)));

    fileName = QDir::homePath().append("/.BLUsage.dat");
    readUsage();

    treeModel = new TreeModel(usageModel.usage);
    ui->treeView->setModel(treeModel);
    ui->treeView->resizeColumnToContents(0);

    if (usageModel.username.isEmpty()) {
        showAccountEditor();
    }

    ui->treeView->setAlternatingRowColors(true);
    ui->accountName->setText(usageModel.name);
    ui->progressBar->setHidden(true);
    showLastUpdate();
}

MainWindow::~MainWindow()
{
    delete treeModel;
    delete networkAccessManager;
    delete ui;
}

void MainWindow::about() {
    QMessageBox::about(this, "BLUsage", "Bangla Lion bandwidth usage viewer.\n\n\tVersion 1.0\n\nAuthor: M. Nasimul Haque\n\nemail: nasim.haque@gmail.com\n\nweb: http://www.nasim.me.uk");
}

void MainWindow::showAccountEditor() {
    AccountDialog(&usageModel, this).exec();
    ui->accountName->setText(usageModel.name);
    writeUsage();
    showLastUpdate();
}

void MainWindow::updateUsage() {
    QUrl url = QUrl(QString("https://care.banglalionwimax.com/User"));
    QString auth = QString("%1:%2").arg(usageModel.username).arg(usageModel.password);

    QNetworkRequest request = QNetworkRequest(url);
    request.setRawHeader("Authorization", QString("Basic " + QByteArray(auth.toLocal8Bit().toBase64())).toLocal8Bit());

    QByteArray postData;
    postData.append("Page=UsrSesHit&");
    postData.append("Title=Session Calls&");
    postData.append(QString("UserID=%1&").arg(usageModel.username));
    postData.append(QString("StartDate=%1&").arg(usageModel.start.toString("dd/MM/yyyy")));
    postData.append(QString("EndDate=%1&").arg(usageModel.end.toString("dd/MM/yyyy")));
    postData.append("Submit=Submit");

    QNetworkReply *reply = networkAccessManager->post(request, postData);
    this->connect(reply, SIGNAL(downloadProgress(qint64,qint64)), SLOT(showProgress(qint64,qint64)));

    ui->updateButton->setEnabled(false);
    ui->progressBar->setHidden(false);
    ui->progressBar->setValue(0);
    ui->statusBar->showMessage("Please wait...");
}

void MainWindow::fetchedUsages(QNetworkReply *reply) {
    ui->updateButton->setEnabled(true);
    ui->progressBar->setHidden(true);
    reply->disconnect(this, SLOT(showProgress(qint64,qint64)));

    QString title;
    QString message;

    switch(reply->error()) {
    case QNetworkReply::AuthenticationRequiredError:
    case QNetworkReply::ContentAccessDenied:
        title = QString("Authentication Error");
        message = QString("Please check your account credentials.");
        break;
    case QNetworkReply::NoError:
        if (!usageModel.parse(QString(reply->readAll()))) {
            title = "Parsing error";
            message = usageModel.errorString();
        }
        else {
            treeModel->deleteLater();
            treeModel = new TreeModel(usageModel.usage);
            ui->treeView->setModel(treeModel);
            ui->treeView->resizeColumnToContents(0);
            showLastUpdate();
            writeUsage();
        }
        return;
    default:
        title = QString("An error occured");
        message = reply->errorString().append(".\nPlease check your internet connection.");
    }

    QMessageBox::critical(this, title, message, QString("OK"));
    reply->deleteLater();
}

void MainWindow::allowConnection(QNetworkReply *reply) {
    reply->ignoreSslErrors();
}

void MainWindow::showProgress(qint64 received, qint64 total) {
    if (total <= 0)
        total = 100000;
    ui->progressBar->setValue(float(received)/float(total)*100.);
}

void MainWindow::showLastUpdate() {
    if (usageModel.lastUpdate.isNull()) {
        ui->statusBar->showMessage("Ready");
    }
    else {
        ui->totalLabel->setText(QString("Totals usage: <b>%1</b>").arg(usageModel.smartBytes(usageModel.totalKB)));
        if (usageModel.capKB > 0) {
            ui->remainingLabel->setText(QString("Remaining: <b>%1</b>").arg(usageModel.smartBytes(usageModel.capKB - usageModel.totalKB)));
        }
        else {
            ui->remainingLabel->setText("Remaining: <b>Unlimited</b>");
        }
        ui->statusBar->showMessage(usageModel.lastUpdate.toString("dd MMM, yyyy HH:mm").prepend("Last updated on: "));
    }
}

void MainWindow::readUsage() {
    QFile file(fileName);
    if(!file.open(QIODevice::ReadOnly))
        return;
    QDataStream in(&file);
    in >> usageModel;
    file.close();
}

void MainWindow::writeUsage() {
    QFile file(fileName);
    if(!file.open(QIODevice::WriteOnly))
        return;
    QDataStream out(&file);
    out << usageModel;
    file.close();
}

File: mainwindow.py
import os
import urllib
import datetime

from PySide.QtCore import QUrl, QSettings
from PySide.QtGui import QMainWindow, QMessageBox, QIcon
from PySide.QtNetwork import (QNetworkAccessManager, QNetworkRequest,
                              QNetworkReply)

from blusage import BLUsage
from treemodel import TreeModel

from accountdialog import AccountDialog
from mainwindow_UI import Ui_MainWindow

class MainWindow(QMainWindow, Ui_MainWindow):

    def __init__(self, parent=None):
        super(MainWindow, self).__init__(parent)
        self.setupUi(self)
        self.setWindowIcon(QIcon(':/blusage.png'))

        self.action_Account.triggered.connect(self.show_account_editor)
        self.action_Update.triggered.connect(self.update_usage)
        self.action_About.triggered.connect(self.about)

        self.network_access_manager = QNetworkAccessManager()
        self.network_access_manager.finished.connect(self.fetched_usages)
        self.network_access_manager.sslErrors.connect(self.allow_connection)

        self.settings = QSettings('Nasim', 'BLUsage')
        self.file_name = os.path.expanduser('~/.BLUsage.dat')

        self.usage_model = BLUsage()
        self.read_usage()
        self.tree_model = TreeModel(self.usage_model.usage)

        self.treeView.setModel(self.tree_model)
        self.treeView.resizeColumnToContents(0)

        if not self.usage_model.username:
            self.show_account_editor()

        self.treeView.setAlternatingRowColors(True)
        self.accountName.setText(self.usage_model.name)
        self.progressBar.setHidden(True)
        self.show_lastupdate()

    def show_account_editor(self):
        AccountDialog(self.usage_model, self).exec_()
        self.accountName.setText(self.usage_model.name)
        self.show_lastupdate()
        self.write_usage()

    def about(self):
        QMessageBox.about(self, "BLUsage", "Bangla Lion bandwidth usage viewer.\n\n\tVersion 1.0\n\nAuthor: M. Nasimul Haque\n\nemail: nasim.haque@gmail.com\n\nWeb: http://www.nasim.me.uk")

    def update_usage(self):
        if not all([self.usage_model.username, self.usage_model.password]):
            QMessageBox.critical(self, 'No account',
                                 'Please enter your account details first.')
            return

        url = QUrl("https://%s:%s@care.banglalionwimax.com/User" %
                   (self.usage_model.username, self.usage_model.password))
        post_data = urllib.urlencode({
            'Page': 'UsrSesHit',
            'Title': 'Session Calls',
            'UserID': self.usage_model.username,
            'StartDate': self.usage_model.start.toString("dd/MM/yyyy"),
            'EndDate': self.usage_model.end.toString("dd/MM/yyyy"),
            'Submit': 'Submit',
        })

        request = QNetworkRequest(url)
        self.network_access_manager.post(request, post_data)

        self.updateButton.setEnabled(False)
        self.progressBar.setHidden(False)
        self.progressBar.setValue(0)
        self.statusBar.showMessage("Please wait...")

    def fetched_usages(self, reply):
        self.updateButton.setEnabled(True)
        self.progressBar.setHidden(True)

        if reply.error() == QNetworkReply.NoError:
            if not self.usage_model.parse(str(reply.readAll())):
                title = "Parsing error"
                message = self.usage_model.error
            else:
                title = None
                self.tree_model.deleteLater()
                self.tree_model = TreeModel(self.usage_model.usage)
                self.treeView.setModel(self.tree_model)
                self.treeView.resizeColumnToContents(0)
                self.show_lastupdate()
                self.write_usage()
        elif reply.error() in [QNetworkReply.AuthenticationRequiredError,
                               QNetworkReply.ContentAccessDenied]:
            title = "Authentication error"
            message = "Please check your account credentials."
        else:
            title = "An error occured"
            message = reply.errorString() \
                    + ".\nPlease check your internet connection."

        if title:
            QMessageBox.critical(self, title, message)
        reply.deleteLater()

    def allow_connection(self, reply):
        reply.ignoreSslErrors()

    def read_usage(self):
        self.usage_model.name = self.settings.value('name')
        self.usage_model.username = self.settings.value('username')
        self.usage_model.password = self.settings.value('password')
        self.usage_model.last_update = self.settings.value('last_update')
        self.usage_model.capKB = int(self.settings.value('cap') or 0)
        self.usage_model.totalKB = int(self.settings.value('total') or 0)

        start = self.settings.value('start')
        if not start:
            today = datetime.date.today()
            start = datetime.date(year=today.year, month=today.month, day=1)
        self.usage_model.start = start
        end = self.settings.value('end')
        if not end:
            end = start + datetime.timedelta(days=29)
        self.usage_model.end = end

    def write_usage(self):
        self.settings.setValue('name', self.usage_model.name)
        self.settings.setValue('username', self.usage_model.username)
        self.settings.setValue('password', self.usage_model.password)
        self.settings.setValue('start', self.usage_model.start)
        self.settings.setValue('end', self.usage_model.end)
        self.settings.setValue('last_update', self.usage_model.last_update)
        self.settings.setValue('cap', self.usage_model.capKB)
        self.settings.setValue('total', self.usage_model.totalKB)

    def show_lastupdate(self):
        if not self.usage_model.last_update:
            self.statusBar.showMessage("Ready")
        else:
            self.totalLabel.setText("Totals usage: <b>%s</b>" %
                    (self.usage_model.smart_bytes(self.usage_model.totalKB),))
            remaining = self.usage_model.remaining()
            if isinstance(remaining, int):
                self.remainingLabel.setText("Remaining: <b>%s</b>" %
                        (self.usage_model.smart_bytes(remaining),))
            else:
                self.remainingLabel.setText("Remaining: <b>Unlimited</b>")

            self.statusBar.showMessage("Last updated on: " +
                    self.usage_model.last_update.strftime("%d %b %Y, %H:%M"))

import resources

